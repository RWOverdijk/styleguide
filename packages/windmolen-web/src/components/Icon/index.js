// @flow
import type { Node } from 'react';
import React from 'react';
import styled from 'styled-components';
import Span from '../Span';
import SpritesheetSVG from '../../assets/icons/sprite-sheet.svg';
import SpritesheetPNG from '../../assets/icons/sprite-sheet.png';
import { colors } from '../../globals';

type Props = {
  name: string,
  fontSize: string,
  variant?: number,
};

const spritesheet = {};

// Define the amount of rows (1-index based).
spritesheet.rows = 5;

// Define the icons from left to right in the sheet using a human-readable name.
spritesheet.names = [
  'facebook',
  'twitter',
  'youtube',
  'instagram',
  'arrow-left',
  'arrow-right',
  'search',
  'thumb-up',
  'app',
  'laadpas',
  'laadpaal',
  'slim-laden',
  'solar-panel',
  'windmill',
  'bio',
  'send',
  'sun',
  'user',
  'phone',
  'app--left',
  'laadpas--left',
  'laadpaal--left',
  'slim-laden--left',
  'leaf',
  'windmill-blades',
  'caret-right',
  'star',
  'play',
  'pause',
  'enter-fullscreen',
  'exit-fullscreen',
  'volume-muted',
  'volume-low',
  'volume-mid',
  'volume-low',
  'wind-indication',
  'star-outline--thick',
  'vdb-logo',
  'wallet--thick',
  'meter--thick',
  'switch--thick',
  'solarpanel--thick',
  'thunderstrike--thick',
  'user--thick',
  'paper--thick',
  'label--thick',
  'bell--thick',
  'lamp--thick',
  'meter-switch--thick',
  'laadpas--thick',
  'laadpaal--thick',
  'app--thick',
  'thunderstrike',
  'wallet',
  'repair',
  'battery',
  'laptop',
  'calendar',
  'notes',
  'phone-thin',
  'gas',
  'apartment',
  'house--tussenwoning',
  'house--hoekwoning',
  'house--2-onder-1-kap',
  'house--vrijstaand',
  'people--one',
  'people--two',
  'people--three',
  'people--four',
  'people--five',
  'electricity-gas',
  'apartment--thick',
  'house--tussenwoning--thick',
  'house--hoekwoning--thick',
  'house--2-onder-1-kap--thick',
  'house--vrijstaand--thick',
  'people--one--thick',
  'people--two--thick',
  'people--three--thick',
  'people--four--thick',
  'people--five--thick',
  'electricity-gas--thick',
  'radio',
  'radio--checked',
  'checkbox',
  'checkbox--checked',
  'checkmark',
];

const StyledSpan = styled(Span)`
  background-image: url('${SpritesheetPNG}');
  background-image: url('${SpritesheetSVG}');
  background-position: -${props => spritesheet.names.indexOf(props.name)}em -${props => props.variant}em;
  background-repeat: no-repeat;
  background-color: ${colors.transparent};
  background-size: ${spritesheet.names.length}em ${spritesheet.rows}em;
  display: inline-block;
  height: 1em;
  width: 1em;
  vertical-align: middle;

  &&& {
    font-size: ${props => props.fontSize};
    line-height: 1;
  }
}`;

const Icon = (props: Props): Node => (
  <StyledSpan {...props} />
);

Icon.defaultProps = {
  fontSize: '150%',
  variant: 0,
};

export {
  spritesheet,
};

export default Icon;
