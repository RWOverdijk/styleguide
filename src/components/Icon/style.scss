.Van-Icon {
  display: inline-block;
  width: 20px;
  height: 20px;
  vertical-align: middle;

  > svg {
    position: relative;
    z-index: 2;
    display: block;
    width: 100%;
    height: 100%;
  }
}
