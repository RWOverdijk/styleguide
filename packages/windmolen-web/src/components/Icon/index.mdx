---
name: Icon
route: /Elements/Icon
menu: Elements
---

import { Playground, PropsTable } from 'docz';
import Icon from './';
import windmill from '../../assets/icons/windmill.svg';

# Icon

## Props
<PropsTable of={Icon} />

## Basic usage
<Playground>
  <Icon source={windmill} />
</Playground>
