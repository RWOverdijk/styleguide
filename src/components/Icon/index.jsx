/* eslint-disable react/no-danger */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import isSVG from 'is-svg';

import './style.scss';

class Icon extends PureComponent {
  render() {
    const { className, source, ...props } = this.props;

    if (!isSVG(source)) {
      return null;
    }

    return (
      <span
        className={classnames('Van-Icon', className)}
        dangerouslySetInnerHTML={{ __html: source }}
        {...props}
      />
    );
  }
}

Icon.defaultProps = {
  className: '',
  source: ''
};

Icon.propTypes = {
  className: PropTypes.string,
  source: PropTypes.string
};

export default Icon;
