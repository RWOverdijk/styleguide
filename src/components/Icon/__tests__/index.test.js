import React from 'react';
import renderer from 'react-test-renderer';
import Icon from '../';

const facebookIcon = '<svg xmlns="http://www.w3.org/2000/svg"><path d="M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z"/></svg>';

describe("<Icon>", () => {
  it("renders correct markup", () => {
    const tree = renderer.create(
      <Icon source={facebookIcon} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders null if source is not a SVG", () => {
    const tree = renderer.create(
      <Icon source={'<img src="my-image-src" />'} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("adds spin animation selector", () => {
    const tree = renderer.create(
      <Icon source={facebookIcon} animation="spin" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("adds pulse animation selector", () => {
    const tree = renderer.create(
      <Icon source={facebookIcon} animation="pulse" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with specific className", () => {
    const tree = renderer.create(
      <Icon source={facebookIcon} className="my-class" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders with specific attributes", () => {
    const tree = renderer.create(
      <Icon source={facebookIcon} data-id="0" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
