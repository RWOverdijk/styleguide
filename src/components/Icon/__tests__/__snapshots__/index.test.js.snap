// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon> adds pulse animation selector 1`] = `
<span
  className="Van-Icon"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"Van-u-animation--pulse\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z\\"/></svg>",
    }
  }
/>
`;

exports[`<Icon> adds spin animation selector 1`] = `
<span
  className="Van-Icon"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"Van-u-animation--spin\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z\\"/></svg>",
    }
  }
/>
`;

exports[`<Icon> renders correct markup 1`] = `
<span
  className="Van-Icon"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z\\"/></svg>",
    }
  }
/>
`;

exports[`<Icon> renders null if source is not a SVG 1`] = `null`;

exports[`<Icon> renders with specific attributes 1`] = `
<span
  className="Van-Icon"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z\\"/></svg>",
    }
  }
  data-id="0"
/>
`;

exports[`<Icon> renders with specific className 1`] = `
<span
  className="Van-Icon my-class"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M1343 12v264h-157q-86 0-116 36t-30 108v189h293l-39 296h-254v759H734V905H479V609h255V391q0-186 104-288.5T1115 0q147 0 228 12z\\"/></svg>",
    }
  }
/>
`;
